### DISCUSSION TITLE
Issue on connection to Neon cluster

### DISCUSSION THREAD

**Discussion by denysloshkarev** [vercel/community#6592] [Help]
### Summary

We had issues with connecting to our PgSQL Neon cluster today.
It worked well for a while, but after some time, it started not responding to lambda API functions but the database cluster responds well for a local.

After redeploying, it started working well again for a while.

How can we check what exactly is happening there and why we lost connect between Vercel and Neon?

It is separate Neon cluster, not the Vercel Postgres

### Steps to Reproduce

Have no Idea how to reproduce it, all worked well the last two weeks. Issues start only today.

**Comment by versecafe** [comment]
@denysloshkarev Do you have any logs from Vercel or Neon that you could share from when the connection is lost?

**Reply by denysloshkarev** [reply]
The latest request before the lost of connection we have today at
```April 22 10:04:46.15 GMT+03:00 RequestID: svm7j-1713769484902-30d651763a95```

after that we started to see that every request after the
```April 22 10:31:55.78 GMT+03:00 Request ID: xw2mc-1713771108503-4dff21aa950d```

cannot connect to the DB cluster and produce the same error:
```
Failed database connection try:  1
Failed database connection try:  2
Error: Connection terminated due to connection timeout
    at Connection.<anonymous> (/var/task/node_modules/pg/lib/client.js:132:73)
    at Object.onceWrapper (node:events:631:28)
    at Connection.emit (node:events:517:28)
    at Connection.emit (node:domain:489:12)
    at Socket.<anonymous> (/var/task/node_modules/pg/lib/connection.js:63:12)
    at Socket.emit (node:events:529:35)
    at Socket.emit (node:domain:489:12)
    at TCP.<anonymous> (node:net:350:12)
    at TCP.callbackTrampoline (node:internal/async_hooks:128:17)
GET /api/homepage 500 3043 ms
```

We use typeorm and In our reconnector the first we try to load the stale connection and close it if it still open, as I mentioned it works fine for last two weeks after we moved to Neon. And today it is the first day when it happened.
```
const initConnection = async () => {
  // clean up old connection that references outdated hot-reload classes
  try {
    const staleConnection = getConnection();
    await staleConnection.close();
  } catch (error) {
    // no stale connection to clean up
  }

  // wait for new default connection
  return await createConnection(connectionOptions);
};

export const prepareConnection = async () => {
  if (!connectionReadyPromise) {
    connectionReadyPromise = initConnection();
  }
  return connectionReadyPromise;
};

const withDatabase = async (req: NextApiRequest, res: NextApiResponse, next: any) => {
  try {
    for (let i = 1; i <= 3; i++) {
      try {
        req.dbConnection = await prepareConnection();
        break;
      } catch (error) {
        if (i >= 3) {
          throw error;
        }
        console.log('Failed database connection try: ', i);
        Sentry.captureMessage(`Failed database connection try ${i}`);
      }
    }
    return next();
  } catch (error) {
    console.error(error);
    Sentry.captureException(error);
    return res.status(500).send('Database connection error');
  }
};
```


We just redeployed Vercel build at `22 Apr 2024, 10:44:38 EEST` and everything started to works again up to the latest good response
```April 22 11:13:08.41 GMT+03:00 Request ID: gxqwn-1713773587248-0f1b5c884b0c```

and again got the issues from
```April 22 12:00:23.34 GMT+03:00 Request ID: rbntk-1713776415401-10adfc5d52e0```
till
```April 22 14:12:32.10 GMT+03:00 Request ID: wj8z6-1713784351789-21eb5161a407```

and after that at
```April 22 14:45:41.24 GMT+03:00 Request ID: j9zfh-1713786336137-e70cc70925a2```
the connections came back, but we not saw it and redeployed build again at `22 Apr 2024, 14:41:17 EEST`


I found the old thread with the similar issues I think
https://github.com/orgs/vercel/discussions/4539
and there is the idea that Vercel freezing the functions with opened connections to DB, is it true?

Because I not saw stale connections in the Neon dashboard so for me it looks like the Vercel's host just loose connections to Neon cluster.

**Reply by amyegan** [reply]
I'm not sure exactly what is causing this issue, but it does seem like it could be related to cold starts or connection pooling. A common solution for some of these problems is to limit the connection count to 1, as suggested in https://github.com/vercel/community/discussions/4539#discussioncomment-9132878

Here are a few other resources that may help connection latency is the issue.
- Neon - [Connection latency and timeouts](https://neon.tech/docs/connect/connection-latency)
- Vercel - [Connection Pooling with Serverless Functions](https://vercel.com/guides/connection-pooling-with-serverless-functions)
- Vercel - [Using Databases with Vercel](https://vercel.com/guides/using-databases-with-vercel)
